demo/opponent.c

```diff
@@ -1,15 +1,32 @@
 #include "shared.h"
 
 /* TODO: Duplicate ndraw in test.c */
-static void ndraw(mp_int *a, const char *name)
+static void ndraw(mp_int * a, const char *name)
 {
-   char buf[16000]; /* TODO: buffer might overflow! */
+   char *buf;
+   int size, size16;
+
+   mp_radix_size(a, 10, &size);
+   mp_radix_size(a, 16, &size16);
+   if (size16 > size) {
+      size = size16;
+   }
+
+   buf = malloc((size_t) size);
```

Comment:
What is a bit bad about this ndraw - it always allocates and ndraw is called quite often in mtest-vs-test.
However in this case the malloc requests should be satisfied by the malloc thread local cache, which is usually there to work around stupid allocation schemes like this. Now @czurnieden will come and find some historical systems where that's not the case :D Maybe he also comes up with a better solution? Allocate the buffer only once and make it global again? Or use alloca :hankey:?

Reply:
As far as I understand that old code, `ndraw` and `draw` are only called in case of an error. O would have suggested a rough approximation of `ceil(log_2(mp_int)/log_2(10))+angst_allowance` to check if   `buf[16000]` is big enough.

> czurnieden will come and find some historical systems where that's not the case

Of course I will! ;-)

Ah, I'm too late. OK, off to #169

Reply:
I have to look more carefully. You are right. It happens only in case of failure. No problem with the malloced ndraw in the new PR #169.